apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/codegen/CodegenLayout.kt

```diff
@@ -17,13 +20,62 @@ import com.apollographql.apollo3.compiler.singularize
  *
  * Inputs should always be GraphQL identifiers and outputs are valid Kotlin/Java identifiers.
  */
-abstract class CodegenLayout(
+internal abstract class CodegenLayout(
+    private val ir: Ir,
     private val packageNameGenerator: PackageNameGenerator,
     private val schemaPackageName: String,
     private val useSemanticNaming: Boolean,
     private val useSchemaPackageNameForFragments: Boolean,
-    private val nameToClassName: (String) -> String,
 ) {
+  private val schemaTypeToClassName: Map<IrSchemaType, String> = computeClassNames()
+
+  private fun computeClassNames(): Map<IrSchemaType, String> {
+    val schemaTypeToClassName = mutableMapOf<IrSchemaType, String>()
+    val allTypes: List<IrSchemaType> = ir.customScalars + ir.objects + ir.enums + ir.interfaces + ir.inputObjects + ir.unions
+    val usedNames = mutableSetOf<String>()
+    for (type in allTypes) {
+      val targetName = type.targetName
+      val className = if (targetName != null) {
+        // If a targetName was specified, honor it verbatim
+        targetName
```

Comment:
targetName certainly needs to be added to `usedNames` here

Reply:
Unless this is caught by validation earlier 🤔 ? In which case add a comment?

Reply:
I _think_ the algorithm should be like so:

1. for all types without a `targetName`, assign a unique (possibly suffixed) name
2. for all types with a `targetName`, assert they are using amongst all types with a `targetName` + the names assigned in 1.

I believe the current algorithm fails in this case:

```graphql
type URL @targetName(newName: "Url2")
type Url 
type url
```

Reply:
Good point. All right, assert it is!